File: src/main/java/gg/mineral/practice/inventory/menus/CategorizedEloMenu.java
package gg.mineral.practice.inventory.menus;

import gg.mineral.practice.category.Category;
import gg.mineral.practice.entity.ProfileData;
import gg.mineral.practice.gametype.Gametype;
import gg.mineral.practice.inventory.ClickCancelled;
import gg.mineral.practice.queue.Queuetype;
import gg.mineral.practice.queue.QueuetypeMenuEntry;
import gg.mineral.practice.util.messages.CC;
import it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap;

@ClickCancelled(true)
public class CategorizedEloMenu extends EloMenu {
    private final Category category;

    public CategorizedEloMenu(ProfileData arg, Queuetype queuetype, Category category) {
        super();
        this.category = category;
        this.arg = arg;
        this.queuetype = queuetype;
        this.menuEntries = getMenuEntries();
    }

    @Override
    protected Object2IntLinkedOpenHashMap<QueuetypeMenuEntry> getMenuEntries() {
        Object2IntLinkedOpenHashMap<QueuetypeMenuEntry> menuEntries = new Object2IntLinkedOpenHashMap<>();
        category.getGametypes().forEach(gametype -> {
            if (gametype.isInCategory())
                menuEntries.put(gametype, queuetype.getMenuEntries().getInt(gametype));
        });
        return menuEntries;
    }

    @Override
    protected boolean shouldSkip(QueuetypeMenuEntry menuEntry) {
        return menuEntry instanceof Gametype gametype && !gametype.isInCategory();
    }

    @Override
    public String getTitle() {
        return CC.BLUE + category.getDisplayName();
    }

    @Override
    public boolean shouldUpdate() {
        return true;
    }
}

File: src/main/kotlin/gg/mineral/practice/inventory/menus/CategorizedEloMenu.kt
package gg.mineral.practice.inventory.menus

import gg.mineral.practice.category.Category
import gg.mineral.practice.entity.ProfileData
import gg.mineral.practice.gametype.Gametype
import gg.mineral.practice.inventory.ClickCancelled
import gg.mineral.practice.inventory.Menu
import gg.mineral.practice.managers.GametypeManager
import gg.mineral.practice.queue.Queuetype
import gg.mineral.practice.queue.QueuetypeMenuEntry
import gg.mineral.practice.util.items.ItemStacks
import gg.mineral.practice.util.messages.CC
import it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap

@ClickCancelled(true)
class CategorizedEloMenu(
    arg: ProfileData,
    queuetype: Queuetype,
    private val category: Category,
    private val prevMenu: Menu
) :
    EloMenu(arg, queuetype) {
    override val menuEntries: Object2IntLinkedOpenHashMap<QueuetypeMenuEntry> by lazy {
        val value = Object2IntLinkedOpenHashMap<QueuetypeMenuEntry>()
        category.gametypes.mapNotNull { GametypeManager.gametypes[it] }
            .forEach { if (it.inCategory) value.put(it, queuetype.menuEntries.getInt(it)) }
        value
    }

    override fun shouldSkip(menuEntry: QueuetypeMenuEntry) = menuEntry is Gametype && !menuEntry.inCategory

    override fun update() {
        super.update()

        addOnNextRow(
            13, ItemStacks.BACK
        ) { viewer.openMenu(prevMenu) }
    }

    override val title: String
        get() = CC.BLUE + category.displayName

    override fun shouldUpdate() = true
}
